File: projects/10-wildfire-forest-monitoring/fog/publisher.js
"use strict";

const { SQSClient, GetQueueUrlCommand, SendMessageCommand, SendMessageBatchCommand } = require("@aws-sdk/client-sqs");

// No class (03's QueueGateway) and no closure-factory holding client/queueUrl
// in a returned object (06's createPublisher). Just a plain function that
// takes the SQS client as a parameter on every call. The only piece of
// state worth keeping across calls is the resolved queue URL lookup itself
// (repeating GetQueueUrlCommand on every publish would be wasteful), so that
// one lookup is memoized behind a module-level Promise cache keyed by queue
// name -- deliberately not a stateful object/class, just a plain Map guarding
// against duplicate concurrent lookups for the same name.
const queueUrlCache = new Map();

function resolveQueueUrl(client, queueName, retries = 30, delayMs = 2000) {
  if (queueUrlCache.has(queueName)) return queueUrlCache.get(queueName);

  const lookup = (async () => {
    for (let attempt = 1; attempt <= retries; attempt++) {
      try {
        const { QueueUrl } = await client.send(new GetQueueUrlCommand({ QueueName: queueName }));
        return QueueUrl;
      } catch (err) {
        if (attempt === retries) {
          queueUrlCache.delete(queueName);
          throw new Error(`queue ${queueName} not reachable: ${err.message}`);
        }
        await new Promise((resolve) => setTimeout(resolve, delayMs));
      }
    }
  })();

  queueUrlCache.set(queueName, lookup);
  return lookup;
}

function buildClient(endpoint, region) {
  // The static test/test pair is a LocalStack convention, so it is only
  // attached when an explicit emulator endpoint is configured. Without one
  // (a real deployment) the SDK's default chain supplies the EC2 instance
  // profile's credentials instead.
  const config = { region };
  if (endpoint) {
    config.endpoint = endpoint;
    config.credentials = { accessKeyId: "test", secretAccessKey: "test" };
  }
  return new SQSClient(config);
}

// The single exported operation: publish one message. Every call receives
// the client explicitly rather than reaching into module-scope state, which
// is what makes this "just a function" rather than an object wrapping a
// connection.
async function publish(sqsClient, queueName, payload, retries, delayMs) {
  const queueUrl = await resolveQueueUrl(sqsClient, queueName, retries, delayMs);
  await sqsClient.send(new SendMessageCommand({
    QueueUrl: queueUrl,
    MessageBody: JSON.stringify(payload),
  }));
}

// SendMessageBatch accepts at most ten entries per call, so a whole flush
// window's aggregates go out in ceil(n/10) batch calls rather than n
// individual SendMessage calls. Same plain-function style as publish():
// the client arrives as a parameter, nothing is wrapped in an object.
const BATCH_LIMIT = 10;

async function publishBatch(sqsClient, queueName, payloads, retries, delayMs) {
  if (!payloads.length) return 0;
  const queueUrl = await resolveQueueUrl(sqsClient, queueName, retries, delayMs);
  let calls = 0;
  for (let start = 0; start < payloads.length; start += BATCH_LIMIT) {
    const entries = payloads.slice(start, start + BATCH_LIMIT).map((payload, offset) => ({
      Id: String(start + offset),
      MessageBody: JSON.stringify(payload),
    }));
    await sqsClient.send(new SendMessageBatchCommand({ QueueUrl: queueUrl, Entries: entries }));
    calls += 1;
  }
  return calls;
}

function clearQueueUrlCache() {
  queueUrlCache.clear();
}

module.exports = { publish, publishBatch, resolveQueueUrl, buildClient, clearQueueUrlCache };

File: projects/12-smart-building-energy/fog/publisher.py
"""SQS publishing as a pair of lru_cache-memoized functions wrapping a plain boto3 client; a failing call is never cached, so _queue_url keeps retrying until the queue exists then stays resolved."""

import functools
import json

import boto3


@functools.lru_cache(maxsize=None)
def _client(endpoint_url, region):
    return boto3.client("sqs", endpoint_url=endpoint_url, region_name=region)


@functools.lru_cache(maxsize=None)
def _queue_url(endpoint_url, region, queue_name):
    return _client(endpoint_url, region).get_queue_url(QueueName=queue_name)["QueueUrl"]


def publish(endpoint_url, region, queue_name, message):
    url = _queue_url(endpoint_url, region, queue_name)
    _client(endpoint_url, region).send_message(QueueUrl=url, MessageBody=json.dumps(message))


def publish_batch(endpoint_url, region, queue_name, messages):
    """Ship a whole flush window in as few SendMessageBatch calls as possible, chunked at the API's hard limit of 10 entries per call."""
    if not messages:
        return
    url = _queue_url(endpoint_url, region, queue_name)
    client = _client(endpoint_url, region)
    for offset in range(0, len(messages), 10):
        chunk = messages[offset:offset + 10]
        entries = [{"Id": str(i), "MessageBody": json.dumps(msg)} for i, msg in enumerate(chunk)]
        client.send_message_batch(QueueUrl=url, Entries=entries)


def reset_cache():
    """Test-only escape hatch: clear the memoized client/queue-url state between tests that point at different fakes."""
    _client.cache_clear()
    _queue_url.cache_clear()
